File: test/domain/find_communes_with_waiting_test.dart
import 'package:coup_laine/domain/models/animal_count.dart';
import 'package:coup_laine/domain/models/client.dart';
import 'package:coup_laine/domain/models/coordinates.dart';
import 'package:coup_laine/domain/use_cases/client_status.dart';
import 'package:coup_laine/domain/use_cases/find_communes_with_waiting.dart';
import 'package:flutter_test/flutter_test.dart';

Client _c(int id, String city, {bool needsRecompute = false}) => Client(
      id: id,
      name: 'C$id',
      addressLabel: 'a',
      postcode: '00000',
      city: city,
      coordinates: const Coordinates(lat: 48, lon: -3),
      animals: const [AnimalCount(categoryId: 1, count: 5)],
      isWaiting: true,
      needsDistanceRecompute: needsRecompute,
    );

void main() {
  test('groups waiting clients by city and sorts alphabetically', () {
    final clients = [
      _c(1, 'Quimper'),
      _c(2, 'Carhaix'),
      _c(3, 'Quimper'),
      _c(4, 'Brest'),
    ];
    final statuses = {
      1: ClientStatus.waiting,
      2: ClientStatus.waiting,
      3: ClientStatus.waiting,
      4: ClientStatus.waiting,
    };
    final result = const FindCommunesWithWaiting().call(
      clients: clients,
      statusByClientId: statuses,
    );
    expect(result.map((e) => e.name), ['Brest', 'Carhaix', 'Quimper']);
    expect(result.map((e) => e.count), [1, 1, 2]);
  });

  test('excludes non-waiting clients', () {
    final clients = [_c(1, 'Quimper'), _c(2, 'Quimper')];
    final statuses = {
      1: ClientStatus.waiting,
      2: ClientStatus.scheduled,
    };
    final result = const FindCommunesWithWaiting().call(
      clients: clients,
      statusByClientId: statuses,
    );
    expect(result, [(name: 'Quimper', count: 1)]);
  });

  test('excludes clients with needsDistanceRecompute', () {
    final clients = [
      _c(1, 'Quimper'),
      _c(2, 'Quimper', needsRecompute: true),
    ];
    final statuses = {
      1: ClientStatus.waiting,
      2: ClientStatus.waiting,
    };
    final result = const FindCommunesWithWaiting().call(
      clients: clients,
      statusByClientId: statuses,
    );
    expect(result, [(name: 'Quimper', count: 1)]);
  });

  test('omits communes whose only client was filtered out', () {
    final clients = [_c(1, 'Carhaix', needsRecompute: true)];
    final statuses = {1: ClientStatus.waiting};
    final result = const FindCommunesWithWaiting().call(
      clients: clients,
      statusByClientId: statuses,
    );
    expect(result, isEmpty);
  });
}

File: tests/domain/use-cases/find-communes-with-waiting.test.ts
import { describe, it, expect } from 'vitest';
import { findCommunesWithWaiting } from '@/domain/use-cases/find-communes-with-waiting';

describe('findCommunesWithWaiting', () => {
  it('groups waiting clients by city, sorted by count desc then name asc', () => {
    const r = findCommunesWithWaiting([
      { id: '1', isWaiting: true, addressCity: 'Quimper' },
      { id: '2', isWaiting: true, addressCity: 'Brest' },
      { id: '3', isWaiting: true, addressCity: 'Quimper' },
      { id: '4', isWaiting: false, addressCity: 'Vannes' },
      { id: '5', isWaiting: true, addressCity: 'Brest' },
      { id: '6', isWaiting: true, addressCity: null },
    ]);
    expect(r).toEqual([
      { city: 'Brest', count: 2 },
      { city: 'Quimper', count: 2 },
    ]);
  });
});
